### DISCUSSION TITLE
tutorials/lift-off-part3/06-query-building-in-apollo-sandbox

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#49] [course feedback]
# tutorials/lift-off-part3/06-query-building-in-apollo-sandbox

https://www.apollographql.com/tutorials/lift-off-part3/06-query-building-in-apollo-sandbox

**Comment by brandonAlder** [comment]
Why do I get an error rather than the desired response when I "*Select all fields recursively*"?  I have tried copy/pasting from the ***final/src*** directory the following files: **index.js**, **resolvers.js**, & **schema.js**.   Additionally I have copied/pasted the request & the full query.

I've noticed the issue is with the track.modules.***id*** field.  If it is omitted, the query/response work as expected.  

There seems to be an issue with the argument's ***id*** clashing with track.modules.id.

Here is the error message I get when I "*Select all fields recursively*".
***"Int cannot represent non-integer value: \"l_0\"",***

**Reply by mabuyo** [reply]
Hey @brandonAlder ! I can't seem to reproduce your error. From the repo, I ran the `final/server` directory with no issues. I also pasted in the full query. I also tried the "select all fields recursively" approach and received the correct data back.

Can you provide the full error? Is it pointing to a specific line in the codebase? 

`l_0` is an id of a module, so I wonder if there is something in the schema that defines it as an `Int` and not an `ID` type.

**Comment by meszarosjeno** [comment]
This is a really nice course so far, but I noticed a minor issue here. In the lesson's last task the query should be named `getMission` (camel case) as the description suggests, but the "checker" only accepts it as `GetMission` (pascal case). Sorry if I'm wrong, keep up the good work! 😁

**Reply by mabuyo** [reply]
Good catch, we should be using PascalCase for consistency. We'll update this code challenge, thanks for letting us know!

**Comment by neogeek13** [comment]
Did I miss adding 'description' somehow? For the step noting  "The full query should look like this:" I don't have 'description' in my Explorer.

I see 'description' in all the server and client files noted exactly match the equivalent final folders EXCEPT client/src/pages/track.tsx (description IS included on line 27 of final/client/src/pages/track.tsx but my version doesn't have it). I didn't think we'd touched any client files in this course so far but maybe I missed a codegen (on the client side) instruction? I pulled the code repo (TS) vs. using what I've built from the first 2 courses so perhaps my skipping over the warning section in the first lesson of this course accounts for the discrepancy?

**Reply by neogeek13** [reply]
After running, I don't see it and, further, I just don't have src/client/pages/track.tsx AT ALL. I have trackS.tsx which appears to match what's in the final folder but I don't have 'track.tsx' which is where I find the equivalent query. The final file is the only one that contains 
const GET_TRACK = gql
Maybe we just haven't gotten that far in this lesson?

**Reply by neogeek13** [reply]
Bah! nm - it's just out of order. XD Think it's time for a break. I'll look forward to seeing the new file in one of the remaining lessons. :P

**Comment by tvadiep** [comment]
Hi there, is there anyone has faced the `CSRF` problem when run the `GetTrack` query on Graphql Explorer?

**Reply by mabuyo** [reply]
Hey @tvadiep - can you show a screenshot of the exact error you're seeing?

**Comment by itsramiel** [comment]
The following line should say `schenma.ts` instead of `schema.js`: 
`The SDL tab shows you your schema represented in schema definition language. It should look familiar from your own schema.js file!`

**Reply by mabuyo** [reply]
Good catch, thank you! Will update this shortly.

**Comment by yingw787** [comment]
Hello,

I'm on the TypeScript tutorial as of February 14, 2024, and I'm getting an error when I copy/pasted this GraphQL query:

```gql
query GetTrack($trackId: ID!) {
  track(id: $trackId) {
    id
    title

    author {
      id
      name
      photo
    }

    thumbnail
    length
    modulesCount
    numberOfViews

    modules {
      id
      title
      length
    }

  }


}
```

This is the data I got back:

{
  "data": {
    "track": null
  },
  "errors": [
    {
      "message": "Cannot return null for non-nullable field Module.id.",
      "locations": [
        {
          "line": 18,
          "column": 7
        }
      ],
      "path": [
        "track",
        "modules",
        0,
        "id"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "Error: Cannot return null for non-nullable field Module.id.",
          "    at completeValue (/home/yingw787/src/dev/odyssey-lift-off-part3/server/node_modules/graphql/execution/execute.js:594:13)",
          "    at executeField (/home/yingw787/src/dev/odyssey-lift-off-part3/server/node_modules/graphql/execution/execute.js:489:19)",
          "    at executeFields (/home/yingw787/src/dev/odyssey-lift-off-part3/server/node_modules/graphql/execution/execute.js:413:20)",
          "    at completeObjectValue (/home/yingw787/src/dev/odyssey-lift-off-part3/server/node_modules/graphql/execution/execute.js:914:10)",
          "    at completeValue (/home/yingw787/src/dev/odyssey-lift-off-part3/server/node_modules/graphql/execution/execute.js:635:12)",
          "    at completeValue (/home/yingw787/src/dev/odyssey-lift-off-part3/server/node_modules/graphql/execution/execute.js:584:23)",
          "    at /home/yingw787/src/dev/odyssey-lift-off-part3/server/node_modules/graphql/execution/execute.js:696:25",
          "    at Function.from (<anonymous>)",
          "    at completeListValue (/home/yingw787/src/dev/odyssey-lift-off-part3/server/node_modules/graphql/execution/execute.js:676:34)",
          "    at completeValue (/home/yingw787/src/dev/odyssey-lift-off-part3/server/node_modules/graphql/execution/execute.js:607:12)"
        ]
      }
    }
  ]
}

From my `server` directory, when I ran `diff . ../final/server`, this is what I came up with:

```bash
Only in .: node_modules
diff ./README.md ../final/server/README.md
3c3
< The starting point of the `server` code for Odyssey Lift-off III course.
---
> The final state of the `server` code after finishing the Odyssey Lift-off III course.
Common subdirectories: ./src and ../final/server/src
```

I made sure to update my server-side types by running `npm run generate`, so my types are up-to-date. I checked out branch TypeScript from the odyssey-lift-off-part3, commit hash `235d14a893335aa23e6aef7f301b29b21a10af55`. I'm on Ubuntu 22.04 LTS, with Node v21.5.0. Can you guys reproduce?

**Reply by yingw787** [reply]
I pushed up my version of the tutorial to this public GitHub repository: git@github.com:yingw787/odyssey-lift-off-part3.git

Problem still remains, HEAD of main

**Reply by lizhennessy** [reply]
Hi there @yingw787 ! It looks like you are missing the `Track.modules` resolver function in your `resolvers.ts` file.

```
  Track: {
    author: ({ authorId }, _, { dataSources }) => {
      return dataSources.trackAPI.getAuthor(authorId);
    },
    modules: ({id}, _, {dataSources}) => {
      return dataSources.trackAPI.getTrackModules(id);
    },
  },
```

In the previous lesson, we wrote the method to retrieve an individual track's modules (I see that's present in your code!) but near the end we also added this resolver function. Try it out and let me know if the error persists!

**Reply by yingw787** [reply]
It works! Thank you thank you!

**Reply by Irina-Volkhonskaya-Imprivata** [reply]
Hi! I have the similar problem, 

My query:
query Track($trackId: ID!) {
  track(id: $trackId) {
    id
    title
  }
}

variables:
{
  "trackId": "c_0" 
}

and I checked my resolvers.ts file. It's correct.
I get an answer:

{
  "data": {
    "track": null
  },
  "errors": [
    {
      "message": "Cannot return null for non-nullable field Track.id.",
      "locations": [
        {
          "line": 3,
          "column": 5
        }
      ],
      "path": [
        "track",
        "id"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "Error: Cannot return null for non-nullable field Track.id.",
          "    at completeValue (C:\\Users\\irina.volkhonskaya\\OneDrive - Imprivata-Corp\\Desktop\\Apollo\\Apollo3\\odyssey-lift-off-part3\\server\\node_modules\\graphql\\execution\\execute.js:605:13)",
          "    at executeField (C:\\Users\\irina.volkhonskaya\\OneDrive - Imprivata-Corp\\Desktop\\Apollo\\Apollo3\\odyssey-lift-off-part3\\server\\node_modules\\graphql\\execution\\execute.js:500:19)",
          "    at executeFields (C:\\Users\\irina.volkhonskaya\\OneDrive - Imprivata-Corp\\Desktop\\Apollo\\Apollo3\\odyssey-lift-off-part3\\server\\node_modules\\graphql\\execution\\execute.js:414:22)",
          "    at completeObjectValue (C:\\Users\\irina.volkhonskaya\\OneDrive - Imprivata-Corp\\Desktop\\Apollo\\Apollo3\\odyssey-lift-off-part3\\server\\node_modules\\graphql\\execution\\execute.js:925:10)",
          "    at completeValue (C:\\Users\\irina.volkhonskaya\\OneDrive - Imprivata-Corp\\Desktop\\Apollo\\Apollo3\\odyssey-lift-off-part3\\server\\node_modules\\graphql\\execution\\execute.js:646:12)",
          "    at C:\\Users\\irina.volkhonskaya\\OneDrive - Imprivata-Corp\\Desktop\\Apollo\\Apollo3\\odyssey-lift-off-part3\\server\\node_modules\\graphql\\execution\\execute.js:497:9",
          "    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)",
          "    at async Promise.all (index 0)",
          "    at async execute (C:\\Users\\irina.volkhonskaya\\OneDrive - Imprivata-Corp\\Desktop\\Apollo\\Apollo3\\odyssey-lift-off-part3\\server\\node_modules\\@apollo\\server\\dist\\cjs\\requestPipeline.js:229:37)",
          "    at async processGraphQLRequest (C:\\Users\\irina.volkhonskaya\\OneDrive - Imprivata-Corp\\Desktop\\Apollo\\Apollo3\\odyssey-lift-off-part3\\server\\node_modules\\@apollo\\server\\dist\\cjs\\requestPipeline.js:158:32)"
        ]
      }
    }
  ]
}


I can't get where is my problem! Please help!

**Reply by Irina-Volkhonskaya-Imprivata** [reply]
I found it) I missed AWAIT

**Comment by pinug** [comment]
There is very important information missing here. Your resolver argument key in your resolvers.js file must be the same as your schema argument key in schema.js. Otherwise, your codes won't work. I struggled with this for 3 days, posted on community but i deleted after no answer.
